File: recording.ts
/**
 * (c) 2024, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */

/*
    Based on
    https://github.com/microsoft/pxt-microbit/blob/master/libs/audio-recording/recording.ts

    Added:
    - sendToSerial()
*/
/*
    The MIT License (MIT)

    Copyright (c) 2022 Lancaster University

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the "Software"),
    to deal in the Software without restriction, including without limitation
    the rights to use, copy, modify, merge, publish, distribute, sublicense,
    and/or sell copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following conditions:
    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.
*/

/**
 * Functions to operate the v2 on-board microphone and speaker.
 */
//% weight=5 color=#015f85 icon="\uf130" block="Record" advanced=false
namespace record {
    // 

    export enum AudioEvent {
        //% block="starts playing"
        StartedPlaying,
        //% block="stops playing"
        StoppedPlaying,
        //% block="starts recording"
        StartedRecording,
        //% block="stops recording"
        StoppedRecording
    }

    export enum AudioLevels {
        //% block="low"
        Low = 1,
        //% block="medium"
        Medium,
        //% block="high"
        High
    }

    export enum AudioSampleRateScope {
        //% block="everything"
        Everything,
        //% block="playback"
        Playback,
        //% block="recording"
        Recording
    }

    export enum AudioRecordingMode {
        //% block="stopped"
        Stopped,
        //% block="recording"
        Recording,
        //% block="playing"
        Playing
    }

    export enum AudioStatus {
        //% block="playing"
        Playing,
        //% block="recording"
        Recording,
        //% block="stopped"
        Stopped,
        //% block="empty"
        BufferEmpty,
    }

    export enum BlockingState {
        //% block="until done"
        Blocking,
        //% block="in background"
        Nonblocking
    }

    let _recordingPresent: boolean = false;

    function audioNotRecording(): boolean {
        return !audioIsRecording();
    }

    function audioNotPlaying(): boolean {
        return !audioIsPlaying();
    }

    /**
     * Record an audio clip for a maximum of 3 seconds
     */
    //% block="record audio clip $mode"
    //% blockId="record_startRecording"
    //% weight=70
    //% parts="microphone"
    //% help=record/start-recording
    export function startRecording(mode: BlockingState): void {
        music._onStopSound(stopPlayback);
        eraseRecording();
        record();
        if (mode === BlockingState.Blocking) pauseUntil(audioNotRecording);
        _recordingPresent = true;
    }

    /**
     * Play recorded audio
     */
    //% block="play audio clip $mode"
    //% blockId="record_playAudio"
    //% weight=60
    //% parts="microphone"
    //% help=record/play-audio
    export function playAudio(mode: BlockingState): void {
        play();
        if (mode === BlockingState.Blocking) pauseUntil(audioNotPlaying);
    }

    function stopPlayback(): void {
        if (audioIsPlaying()) {
            stop();
        }
    }

    //% shim=record::stop
    export function stopRecording(): void {
    }

    export function eraseRecording(): void {
        _recordingPresent = false;
        erase();
        return
    }

    /**
     * Test what the audio is doing
     */
    //% block="audio is $status"
    //% blockId="record_audioStatus"
    //% parts="microphone"
    //% help=record/audio-status
    export function audioStatus(status: AudioStatus): boolean {
        switch (status) {
            case AudioStatus.Playing:
                return audioIsPlaying();
            case AudioStatus.Recording:
                return audioIsRecording();
            case AudioStatus.Stopped:
                return audioIsStopped();
            case AudioStatus.BufferEmpty:
                return !_recordingPresent;
        }
    }

    /**
     * Change how sensitive the microphone is. This changes the recording quality!
     */
    //% block="set microphone sensitivity to $gain"
    //% blockId="record_setMicGain"
    //% parts="microphone"
    //% weight=30
    //% help=record/set-mic-gain
    export function setMicGain(gain: AudioLevels): void {
        switch (gain) {
            case AudioLevels.Low:
                setMicrophoneGain(0.079);
                break;
            case AudioLevels.Medium:
                setMicrophoneGain(0.2);
                break;
            case AudioLevels.High:
                setMicrophoneGain(1.0);
                break;
        }
    }

    /**
     * Set the sample frequency for recording, playback, or both (default)
     * 
     * @param hz The sample frequency, in Hz
     */
    //% block="set sample rate to $hz || for $scope"
    //% blockId="record_setSampleRate"
    //% hz.min=1000 hz.max=22000 hz.defl=11000
    //% expandableArgumentMode="enabled"
    //% parts="microphone"
    //% weight=40
    //% help=record/set-sample-rate
    export function setSampleRate(hz: number, scope?: AudioSampleRateScope): void {
        switch (scope) {
            case AudioSampleRateScope.Playback:
                setOutputSampleRate(hz);
                break;
            case AudioSampleRateScope.Recording:
                setInputSampleRate(hz);
                break;
            case AudioSampleRateScope.Everything:
            default:
                setBothSamples(hz);
                break;
        }
    }

    /**
     * Set the volume for playback
     * 
     * @param volume The volume
     */
    //% block="set playback volume to $volume"
    //% blockId="record_setPlaybackVolume"
    //% volume.min=0 volume.max=1000 volume.defl=75
    //% parts="microphone"
    //% weight=30
    //% help=record/set-volume
    //% shim=record::setPlaybackVolume
    export function setPlaybackVolume(volume: number): void {
        return;
    }

    /**
     * Send recorded audio clip data to serial
     */
    //% block="send audio clip to serial"
	//% blockId="record_sendToSerial"
	//% weight=20
	export function sendToSerial(mode: BlockingState): void {
		send();
		if (mode === BlockingState.Blocking) pauseUntil(notSendingToSerial);
	}

	/**
	 * Test whether sending to serial
     */
    //% block="sending to serial"
	//% blockId="record_sendingToSerial"
    //% shim=record::sendingToSerial
	//% weight=10
    export function sendingToSerial(): boolean {
        return false
	}

    //% shim=record::send
	function send(): void {
		basic.pause(0);
	}

    function notSendingToSerial(): boolean {
		return !sendingToSerial();    
	}

    /**
     * Total length in bytes of the recorded audio clip
     */
    //% block="recording length"
    //% blockId="record_clipLength"
    //% weight=17
    //% shim=record::clipLength
    export function clipLength(): int32 {
        return 0;
    }

    /**
     * Rewind the recorded clip, ready to be read chunk by chunk
     */
    //% block="begin reading recording"
    //% blockId="record_beginExtract"
    //% weight=16
    //% shim=record::beginExtract
    export function beginExtract(): void {
        return;
    }

    /**
     * Read the next chunk of the recorded clip. Returns an empty buffer at the end.
     */
    //% block="next recording chunk"
    //% blockId="record_nextChunk"
    //% weight=15
    //% shim=record::nextChunk
    export function nextChunk(): Buffer {
        return control.createBuffer(0);
    }
}

File: recording.cpp


/**
 * (c) 2024, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */
/*
    Based on
    https://github.com/microsoft/pxt-microbit/blob/master/libs/audio-recording/recording.cpp

    Added:
    - sendToSerial()
*/
/*
    The MIT License (MIT)

    Copyright (c) 2022 Lancaster University

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the "Software"),
    to deal in the Software without restriction, including without limitation
    the rights to use, copy, modify, merge, publish, distribute, sublicense,
    and/or sell copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following conditions:
    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.
*/

#include "pxt.h"
#include "MicroBit.h"

#if MICROBIT_CODAL
#include "StreamRecording.h"
#include "SerialStreamer.h"
#endif

using namespace pxt;

namespace record {

#if MICROBIT_CODAL
// Based on
// https://github.com/lancaster-university/codal-core/blob/master/source/streams/SerialStreamer.cpp
// Copyright (c) 2016 Lancaster University.
class SerialSink : public DataSink
{
    DataSource      &upstream; 
 
public:
    SerialSink( DataSource &source) : upstream( source) {
        source.connect( *this);
        source.dataWanted( DATASTREAM_WANTED);
    }

    int pullRequest() {
        static volatile int pr = 0;
        if ( !pr) {
            pr++;
            while ( pr) {
                send( upstream.pull());
                pr--;
            }
        } else {
            pr++;
        }
        return DEVICE_OK;
    }

    void send( ManagedBuffer buffer) {
        if ( buffer.length() <= 0) return;
        int format = upstream.getFormat();
        int skip = DATASTREAM_FORMAT_BYTES_PER_SAMPLE( format);
        uint8_t *ptr  = &buffer[0]; 
        uint8_t *next = ptr + buffer.length();
        while ( ptr < next) {
            int32_t v = StreamNormalizer::readSample[ format]( ptr);
            Serial::defaultSerial->send( ManagedString( (int) v) + "\n");
            ptr += skip;
        }
    }
};
#endif //MICROBIT_CODAL


#if MICROBIT_CODAL
static StreamRecording *recording = NULL;
static SplitterChannel *splitterChannel = NULL;
static MixerChannel *channel = NULL;
static SerialSink *serialSink = NULL;
#endif


void checkEnv() {
#if MICROBIT_CODAL
    if (recording == NULL) {
        int defaultSampleRate = 11000;
        MicroBitAudio::requestActivation();

        splitterChannel = uBit.audio.splitter->createChannel();
        uBit.audio.mic->setSampleRate( defaultSampleRate );

        recording = new StreamRecording(*splitterChannel);

        channel = uBit.audio.mixer.addChannel(*recording, defaultSampleRate);

        channel->setVolume(75.0);

        serialSink = new SerialSink(*recording);
    }
#endif
}

/**
 * Record an audio clip
 */
//% promise
void record() {
#if MICROBIT_CODAL
    checkEnv();
    recording->recordAsync();
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

/**
 * Play the audio clip that is saved in the buffer
 */
//%
void play() {
#if MICROBIT_CODAL
    checkEnv();
    if ( recording->isPlaying() && recording->downStream != channel) {
        recording->stop();
    }
    recording->connect( *channel);
    recording->playAsync();
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

/**
 * Stop recording
 */
//%
void stop() {
#if MICROBIT_CODAL
    checkEnv();
    recording->stop();
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

/**
 * Clear the buffer
 */
//%
void erase() {
#if MICROBIT_CODAL
    checkEnv();
    recording->erase();
#endif
}

/**
 * Set sensitity of the microphone input
 */
//%
void setMicrophoneGain(float gain) {
#if MICROBIT_CODAL
    uBit.audio.processor->setGain(gain);
#endif
}

/**
 * Get how long the recorded audio clip is
 */
//%
int audioDuration(int sampleRate) {
#if MICROBIT_CODAL
    return recording->duration(sampleRate);
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
    return MICROBIT_NOT_SUPPORTED;
#endif
}

/**
 * Get whether the playback is active
 */
//%
bool audioIsPlaying() {
#if MICROBIT_CODAL
    return recording->isPlaying();
#else
    return false;
#endif
}

/**
 * Get whether the microphone is listening
 */
//%
bool audioIsRecording() {
#if MICROBIT_CODAL
    return recording->isRecording();
#else
    return false;
#endif
}

/**
 * Get whether the board is recording or playing back
 */
//%
bool audioIsStopped() {
#if MICROBIT_CODAL
    return recording->isStopped();
#else
    return false;
#endif
}

/**
 * Change the sample rate of the splitter channel (audio input)
 */
//%
void setInputSampleRate(int sampleRate) {
#if MICROBIT_CODAL
    checkEnv();
    uBit.audio.mic->setSampleRate(sampleRate);
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}


/**
 * Change the sample rate of the mixer channel (audio output)
 */
//%
void setOutputSampleRate(int sampleRate) {
#if MICROBIT_CODAL
    checkEnv();
    channel->setSampleRate(sampleRate);
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

/**
 * Set the sample rate for both input and output
*/
//%
void setBothSamples(int sampleRate) {
#if MICROBIT_CODAL
    setOutputSampleRate(sampleRate);
    uBit.audio.mic->setSampleRate(sampleRate);
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}


//%
void setPlaybackVolume(int volume) {
#if MICROBIT_CODAL
    checkEnv();
    channel->setVolume(volume);
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

//%
void send() {
#if MICROBIT_CODAL
    checkEnv();
    if ( recording->isPlaying() && recording->downStream != serialSink) {
        recording->stop();
    }
    recording->connect( *serialSink);
    recording->playAsync();
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}


//%
bool sendingToSerial() {
#if MICROBIT_CODAL
    checkEnv();
    return recording->isPlaying() && recording->downStream == serialSink;
#else
    return false;
#endif
}

/**
 * Total length in bytes of the recorded audio clip
 */
//%
int clipLength() {
#if MICROBIT_CODAL
    checkEnv();
    return recording->length();
#else
    return 0;
#endif
}

/**
 * Rewind the recorded clip, ready to be read chunk by chunk
 */
//%
void beginExtract() {
#if MICROBIT_CODAL
    checkEnv();

    if (!recording->isStopped())
        recording->stop();

    // checkEnv() leaves serialSink permanently connected downstream (its
    // constructor calls source.connect()). playAsync() calls
    // downStream->pullRequest(), so without disconnecting first, serialSink
    // drains the whole clip to the serial port before the JS side gets to
    // call nextChunk(), which then only ever sees an empty stream.
    recording->disconnect();

    // Rewinds to the start of the clip. With no downstream connected, nothing
    // consumes the stream behind our back: the JS side drives it one chunk at
    // a time through nextChunk().
    recording->playAsync();
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
#endif
}

/**
 * Read the next chunk of the recorded clip. Returns an empty buffer at the end.
 */
//%
Buffer nextChunk() {
#if MICROBIT_CODAL
    checkEnv();
    // One StreamRecording chunk at a time (256 bytes), so we never allocate a
    // copy of the whole clip: at 11kHz a clip can reach 51200 bytes, which is
    // far too big for a single PXT allocation (panic 022).
    ManagedBuffer chunk = recording->pull();
    if (chunk.length() == 0) return mkBuffer(NULL, 0);
    return mkBuffer(chunk.getBytes(), chunk.length());
#else
    target_panic(PANIC_VARIANT_NOT_SUPPORTED);
    return mkBuffer(NULL, 0);
#endif
}

} // namespace record
